### ISSUE TITLE
Aggregator and variable-scoping issue: "Assertion `index.isDefined(*var) && "variable not grounded"' failed."

### ISSUE THREAD

**Issue by cfallin** [souffle-lang/souffle#556]
I'm attempting to run the following Souffle program:

    .decl X(n:number)
    
    X(1). X(2). X(10). X(11).
    
    .decl Out(c:number, n:number)
    
    Out(c, n) :-
      X(n),
      c = count : { X(m), m < n }.
    
    .output Out

When I run `$ souffle test.dl` using the current `master` branch (`7b627374`), I get:

    souffle: AstTranslator.cpp:386: std::unique_ptr<souffle::RamValue> souffle::{anonymous}::translateValue(const souffle::AstArgument*, const souffle::{anonymous}::ValueIndex&): Assertion `index.isDefined(*var) && "variable not grounded"' failed.
    Aborted (core dumped)

If I tweak the code at that line to dump `var` and `index`, I see that `var` is `n` and `index` prints as `(m,{(0,0)})`.

So it appears that the binding for `n`, which should be grounded by the first clause `X(n)` (right?), is somehow not propagating into the aggregator body.

I noticed some recent PRs related to witnesses and aggregator scopes, so perhaps this is a related bug? FWIW, the same issue occurs with a version from before those changes (my own fork based on `fab6f061` of Jan 2).

Thanks for taking the time to look at this!

**Comment by cfallin** [comment]
As a bit of follow-up investigation: it seems that this issue occurs when an aggregator materializes a new relation via `MaterializeAggregationQueriesTransformer`. The synthetic relation appears to be:

    __agg_rel_0(m,n) :- 
       X(m),
       m < n.

Which of course has an ungrounded variable `n`. It seems to me that the transform would need to somehow carry over the set of points/tuples/variable assignments in the context of `Out` via another synthetic relation -- would a transform something like this be appropriate?:

    __agg_rel_0_domain(n) :- X(n).  // Copied from `Out`'s body
    
    __agg_rel_0(m,n) :- 
        __agg_rel_0_domain(n),
       X(m),
       m < n.

(I'm somewhat new to Datalog and I know somewhat little about Souffle's architecture, so forgive me if this is off-base; just one possible idea!)

**Comment by b-scholz** [comment]
I could reproduce the crash. Thanks for the detailed report!  

Unfortunately, aggregates are not tested so well.  We will look into this next week.

**Comment by b-scholz** [comment]
The current implementation of aggregates does not fully cover the semantics of Datalog. Basically, in the current implementation we cannot have constraints in the sub-clause of the aggregate whose variables are not grounded in the sub-clause. 

Instead of fixing the current translation, it might be a better approach to build a new loop-nest for the sub-clause in the loop-nest. Note that at the moment we defer the computation in form of a new table. However, in the example above, this translation strategy fails. 

For the time being, I suggest that you rewrite your Datalog code as follows:
```
.decl X(n:number)

X(1). X(2). X(10). X(11).

.decl Out(c:number, n:number)

Out(c, n) :-
  X(n),
  c = count : { X(n), X(m), m < n }.

.output Out
```

**Comment by mmcgr** [comment]
Closed in favour of #662.
